Ledger sign in

Patrick1904 opened the pull request:
WIP [589](https://github.com/near/near-wallet/issues/589)

vgrichina commented on src/components/accounts/ledger/SignInLedger.js:
```diff
@@ -0,0 +1,53 @@
+            <p><Translate id='signInLedger.one'/></p>
+            <FormButton
+                onClick={handleSignIn}
+                sending={signingIn === 'true'}
```
let's not use `true` as one of the states as it's super confusing, especially given that opposite state seems to be `fail`? Are there any other possible states?  Should we just use latest action's request status from Redux state?

Patrick1904 replied:
@vgrichina good point. I'll use Redux state instead.

kcole16 commented:
@Patrick1904 @vgrichina What's the latest on this PR?

kcole16 commented:
@Patrick1904 Noticing a few issues here:
1. Seed phrase copy isn't working correctly: The words do not have spaces
2. Setting up Ledger is acting strangely when setting it up and deleting keys
  - Ledger is asking you to sign a TX "confirming multiple actions", which I think might be related to the TX deleting the keys (shouldn't require Ledger here, as still have the localStorage key)

Patrick1904 commented:
> 1. Seed phrase copy/paste isn't working correctly: The words do not have spaces

It seems the layout change, specifically 'display: flex' removes the spaces (not all browsers). Not sure if this is a bug since you're not actually copying the phrase, and if we want to allow that, we should put the copy button there, like mobile? Actually copying works fine.


> 2. Setting up Ledger is acting strangely when setting it up and deleting keys

Is the Ledger acting strangely or what part specifically?

> * Ledger is asking you to sign a TX "confirming multiple actions", which I think might be related to the TX deleting the keys (shouldn't require Ledger here, as still have the localStorage key)

@vgrichina any idea why this is happening?

> 1. Looks like the modal isn't popping up to sign a TX with Ledger (though I assume this isn't implemented yet)

Yes have been holding off on that until the basic 'Ledger sign in' works.

----------

The remaining item left (aside from notes above) is 'sign in':
- Get key from Ledger and lookup accountId(s) using the explorer function @marcinbodnar wrote, and then add those to localStorage to sign in the user. @vgrichina are there any new requirements now after you've updated the Ledger app? If not, can @marcinbodnar help with this part?

kcole16 commented:
> It seems the layout change, specifically 'display: flex' removes the spaces (not all browsers). Not sure if this is a bug since you're not actually copying the phrase, and if we want to allow that, we should put the copy button there, like mobile? Actually copying works fine.

We decided to allow users to copy the phrase, but not encourage it on desktop by offering the copy button (though decided to do so on mobile). The reality is most users will want to copy it, and will view the current behavior as a bug. We can take the stance that we don't support this behavior, but I expect we will get persistent bug reports about this. Seems like the best option is to add spaces to the copied text (as we had before).

> Is the Ledger acting strangely or what part specifically?

Ah sorry, the bullet point under #2 is the details. Updated to make more clear.

> Yes have been holding off on that until the basic 'Ledger sign in' works.

Makes sense. I'll split out the issue into "Login/Use in Ledger only mode" and "Sign with Ledger". Fits a bit better with current v0 plans.

Patrick1904 commented:
> We decided to allow users to copy the phrase, but not encourage it on desktop by offering the copy button (though decided to do so on mobile). The reality is most users will want to copy it, and will view the current behavior as a bug. We can take the stance that we don't support this behavior, but I expect we will get persistent bug reports about this. Seems like the best option is to add spaces to the copied text (as we had before).

Yes we can fix so that you can still highlight and copy the text and retain the spaces in Chrome. Seems not a quick fix because it requires writing the layout without flexbox. My POV is still that we should put an actual 'copy' button on desktop, since we're already doing it on mobile. Who has pen and paper anymore :)

vgrichina commented:
> It seems the layout change, specifically 'display: flex' removes the spaces (not all browsers). Not sure if this is a bug

copying is discouraged, but also we cannot prevent it from happening

maybe we actually should also duplicate seed phrase in password field so that you can save it in password manager easily. @Patrick1904  @corwinharrell  @kcole16 wonder what you think on this.

vgrichina commented:
@marcinbodnar @Patrick1904 given I'm back I can continue to work on this. 

> Ledger is asking you to sign a TX "confirming multiple actions", which I think might be related to the TX deleting the keys (shouldn't require Ledger here, as still have the localStorage key)

I think there used to be a bug with Ledger app where it will keep old transaction after signing it once. I think I fixed it in latest updates to Ledger app (in fact this was one of the things delaying direct work on this PR), I can test it all together with latest app to make sure stuff works / fix Ledger app as necessary.

Patrick1904 commented:
> @marcinbodnar @Patrick1904 given I'm back I can continue to work on this.
> 
> > Ledger is asking you to sign a TX "confirming multiple actions", which I think might be related to the TX deleting the keys (shouldn't require Ledger here, as still have the localStorage key)
> 
> I think there used to be a bug with Ledger app where it will keep old transaction after signing it once. I think I fixed it in latest updates to Ledger app (in fact this was one of the things delaying direct work on this PR), I can test it all together with latest app to make sure stuff works / fix Ledger app as necessary.

Sounds good to me. @vgrichina I've rebased the branch so it's up-to-date with master.

Patrick1904 commented:
> > It seems the layout change, specifically 'display: flex' removes the spaces (not all browsers). Not sure if this is a bug
> 
> copying is discouraged, but also we cannot prevent it from happening
> 
> maybe we actually should also duplicate seed phrase in password field so that you can save it in password manager easily. @Patrick1904 @corwinharrell @kcole16 wonder what you think on this.

IMO we should always aim to align the desktop and mobile UX. We offer a 'copy phrase' button on mobile, but we don't on desktop. But we want to make it easy for users to highlight the full phrase on desktop and paste it in a doc etc (and preserve spaces between words). I think we should decide on one policy. Either we allow all users to copy the seed phrase, or we don't. It's not any easier for me to 'write down' the phrase while on desktop, than on mobile.

My perception is that users want to be able to copy the phrase, and will attempt to copy it (by highlighting) even if there's no explicit copy button. My recommendation is therefore that we show the 'copy phrase' button on both desktop and mobile.

vgrichina commented:
> My recommendation is therefore that we show the 'copy phrase' button on both desktop and mobile.

if it's easiest fix I'm ok on this

marcinbodnar commented:
@vgrichina is `createLedgerU2FClient` working for you locally?

vgrichina commented:
@marcinbodnar it used to work, you need to serve website using HTTPS though and use browser with U2F support (e.g. Chrome)

vgrichina commented:
@Patrick1904 @kcole16 should work now. Copy phrase also added back.

kcole16 commented:
Looks like I'm being "logged in" in that the account is showing in the navbar and on the dashboard, but I'm still redirected to the `/create` page and see this error in the console.

`Error loading account: No account.`

I also noticed that the user will need to confirm one `addKey` action per account that has used the Ledger key. I assumed it wasn't working for a while, then realized it was each of the accounts I'd used Ledger for. Will add some copy to `/sign-in-ledger` to give the user some guidance here.

vgrichina commented:
>  but I'm still redirected to the /create page 

This seems like a general pattern where we "remember" some internal pages to redirect to.
@marcinbodnar can you look into this? We only should remember pages which aren't a part of create account / recover / etc flows. Maybe we should only redirect to external pages to simplify this?

> I also noticed that the user will need to confirm one addKey action per account that has used the Ledger key. I assumed it wasn't working for a while, then realized it was each of the accounts I'd used Ledger for. Will add some copy to /sign-in-ledger to give the user some guidance here.

@kcole16 let's have a tracking issue to update this in UI as well. I think we should indicate clearly when we are actually waiting for user to confirm on Ledger and when just some other stuff is going on. So needs some progress indicator reflecting this. /cc @corwinharrell @Patrick1904 

Other than that I think it's best to merge this already and improve in follow-up PRs.

marcinbodnar commented:
@vgrichina We need to add `/sign-in-ledger` to `WALLET_CREATE_NEW_ACCOUNT_FLOW_URLS`. Because now `redirect_url` is cleared when entering `/sign-in-ledger`. But I'm not sure if it would solve the problem. 

I'm finishing for today, but I will check this out tomorrow.

> We only should remember pages which aren't a part of create account / recover / etc flows. Maybe we should only redirect to external pages to simplify this?

It can only happen if for example user enters `https://wallet.nearprotocol.com/create` in a browser and don't have an account or if the user clicks Create Account button while being on any of account flow pages. But we can add a simple condition, that `redirect_url` will not be updated if the URL is one of account flow.

> I also noticed that the user will need to confirm one addKey action per account that has used the Ledger key.

@kcole16 It would be nice if in the middle of the process we could show list of found accounts with additional information that says it is required to confirm for all of them separately in the device. We can also add an icon next to the every account showing if it's the account is already confirmed.
